### ISSUE TITLE
Make SVG export friendly to Inkspace bug

### ISSUE THREAD

**Issue by taylorh140** [typst/typst#1986]
### Description

using Inkscape 1.3 (0e150ed6c4, 2023-07-21)
and prerelease typst (you can't export svgs otherwise) typst 0.7.0 (3fcb5ea7)

I created a image using the following:

````typst
#import "@preview/tbl:0.0.4"

#let rotatex(body,angle) = style(styles => {
  let size = measure(body,styles)
  box(inset:(x: -size.width/2+(size.width*calc.abs(calc.cos(angle))+size.height*calc.abs(calc.sin(angle)))/2,
             y: -size.height/2+(size.height*calc.abs(calc.cos(angle))+size.width*calc.abs(calc.sin(angle)))/2),
             rotate(body,angle))
})

#show: tbl.template.with(allbox: true, tab: "|",scope:("rotatex":rotatex))

```tbl
R Cw(30pt).
#rotatex([Main Scope],-90deg) |\
#sym.dots.v   
\^ | Rtn          
\^ | LR           
\^ | R12          
\^ | R3           
\^ | R2           
\^ | R1           
\^ | R0     
```
````

when imported into inkscape it looks like:
![image](https://github.com/typst/typst/assets/10996543/1d1a1202-0116-4cda-941b-91a78005c6dd)

however moving the <defs> to the top of the svg fixes it.

![image](https://github.com/typst/typst/assets/10996543/21bfb86e-c29a-409b-905d-92808893a31c)

The two svgs are attached. (they both look fine in all browsers I have tried.)

I have not found the issue in the inkscape gitlab server if there is one...
https://gitlab.com/inkscape/inkscape/-/issues

![TC](https://github.com/typst/typst/assets/10996543/62d4ac51-c4f6-4703-9e59-f2a33198f616)
![TCBD](https://github.com/typst/typst/assets/10996543/eb657fd8-1bb9-46d1-9bb7-430bd44b4568)

### Operating system

Windows

### Typst version

- [ ] I am using the latest version of Typst

**Comment by PgBiel** [comment]
Same here by exporting tablex-test to svg and opening in inkscape (to the left, the SVG as exported by typst with `typst c tablex-test.typ tablex-test-{n}.svg` at the tablex 0.0.5 tag; to the right, the same SVG but moving <defs> tags to the top):
![image](https://github.com/typst/typst/assets/9021226/fe17f02e-b4a1-4121-9bef-f976b608f8b1)

Both SVGs appear fine in Firefox.

**Comment by PgBiel** [comment]
Here are the relevant uploads:

<details><summary>Original SVG</summary>
<p>

![tablex-test-01](https://github.com/typst/typst/assets/9021226/bb398fde-7141-4dc8-8c68-e3fd2bb5e211)
</p>
</details> 

<details><summary>Modified/"Fixed" SVG (defs moved to the top)</summary>
<p>

![tablex-test-01 fixed](https://github.com/typst/typst/assets/9021226/1cf1ab52-8a3c-4ff0-a029-df2c92879cd7)

</p>
</details> 

<details><summary>.typ files (export `tablex-test.typ` to `tablex-test-{n}.svg` and take the one with `-01`)</summary>
<p>

[tablex.zip](https://github.com/typst/typst/files/12424309/tablex.zip)

</p>
</details>

**Comment by PgBiel** [comment]
Here are some more interesting results - typst's exported SVGs have a second `<defs>` element which didn't matter in my previous example as it was empty, for clip paths (i.e. boxes and blocks with `clip: true`). I made a sample typst file with the following contents:

```js
#let clippy(x) = box(clip: true, width: 50pt, height: 50pt, x)
#let muhclip(n) = clippy(align(center + horizon, rect(lorem(n))))

#muhclip(5)
#muhclip(40)

#muhclip(100)
```

I, then, exported it to SVG, and made four different SVGs from that, producing the following outputs on inkscape (all of them are normal in Firefox):
![image](https://github.com/typst/typst/assets/9021226/aec66ffe-1de6-4b5f-b0c7-7707dfc9141e)

<details><summary>Top left SVG (original, directly from Typst)</summary>
<p>

![asdf-1](https://github.com/typst/typst/assets/9021226/e49e61c0-5fa8-48d1-9237-dc8ed78312e8)


</p>
</details>

<details><summary>Top right SVG (moved just `&lt;defs id="glyph"&gt;` to the top, but not the one with `id="clip-path"`)</summary>
<p>


![asdf-1 defs-glyph](https://github.com/typst/typst/assets/9021226/82b7e922-9078-4f5f-969d-d22f22748b05)


</p>
</details> 
<details><summary>Bottom left SVG (moved just `&lt;defs id="clip-path"&gt;` to the top, but not the one with `id="glyph"`)</summary>
<p>

![asdf-1 defs-clip](https://github.com/typst/typst/assets/9021226/53017975-0015-4658-ba54-99b0df58ef08)


</p>
</details> 

<details><summary>Bottom right SVG (moved both `&lt;defs id="glyph"&gt;` and the one with  `id="clip-path"` to the top)
</summary>
<p>


![asdf-1 both-defs](https://github.com/typst/typst/assets/9021226/c7f027dd-169d-4ff2-a252-4397fd082359)


</p>
</details> 

It's interesting to note that the one with just the clip-path defs moved to the top has different missing glyphs than the one with both at the bottom.

**Comment by PgBiel** [comment]
This might very well be a bug in inkscape. We could consider an SVG export implementation which just applies the workaround of moving the defs to the top (somehow efficiently).

**Comment by taylorh140** [comment]
I can agree with this. I have yet to identify an existing issue. I might just have to generate a new issue.

**Comment by taylorh140** [comment]
Someone beat me to it (thanks):
https://gitlab.com/inkscape/inbox/-/issues/9119

**Comment by Enter-tainer** [comment]
seems that inkscape people dont have enough bandwidth to fix this. perhaps we can move defs to the top of the svg file to make inkscape happy.

**Comment by Dherse** [comment]
> seems that inkscape people dont have enough bandwidth to fix this. perhaps we can move defs to the top of the svg file to make inkscape happy.

Wouldn't that involve having to allocate a whole new String slowing everything down? We can easily do it but it feels like such a chore and a pointless loss of performance for something that isn't even our fault (saying that, how much of the PDF export code is fixing other people's issues heh?)

**Comment by frozolotl** [comment]
This issue seems to have been fixed since Typst 0.10.0.
